feat: Add lib v2/legacy tabs in studio home

yusuf-musleh opened the pull request:
## Description

This PR adds a new configuration flag that shows/hides tabs in studio home along with some new functionality around to V1 and V2 Libraries.

When the new `LIBRARY_MODE` flag is set to `"mixed"` (default in dev) it will show "Libraries" and "Legacy Libraries" tabs that correspond to v1 and v2 tabs respectively:

<img width="1372" alt="Screenshot 2024-06-03 at 3 47 48 PM" src="https://github.com/openedx/frontend-app-course-authoring/assets/6829768/a7f39296-6fee-4a4b-8d54-338012834683">

<img width="1370" alt="Screenshot 2024-06-03 at 3 48 35 PM" src="https://github.com/openedx/frontend-app-course-authoring/assets/6829768/a39faac8-f4b9-44b7-9064-9d8f89a92ea6">

When the new `LIBRARY_MODE` flag is set to `"v1 only"` (default in production) or `"v2 only"`, only one tab "Libraries" is shown and only the respective libraries are fetched when the tab is clicked.

In addition to the above changes, the URL/route now updates when clicking on the tabs, and navigating to it directly would open up that tab as well as a new placeholder page that you will be redirected to when clicking on a v2 library if the library authoring MFE is not enabled:

<img width="1372" alt="Screenshot 2024-06-03 at 3 51 16 PM" src="https://github.com/openedx/frontend-app-course-authoring/assets/6829768/bfa6339f-8b47-48c3-8264-5a174989798c">

## Supporting information

Related Tickets:
- Depends on: https://github.com/openedx/edx-platform/pull/34879
- Related Issue: https://github.com/openedx/frontend-app-course-authoring/issues/1033

## Testing instructions

1. Run this PR's branch in your local env
1. Make sure you have the https://github.com/openedx/edx-platform/pull/34879/ running locally
1. Make sure you have the `studio.library_authoring_mfe` waffle flag enabled in the 
1. The "Libraries" and "Legacy Libraries" tabs appear in the studio home page
1. When clicking on the "Libraries" shows:
    1. v2 libraries are shown
    1. They should be paginated (you should see "Show x out of y", to further confirm pagination is working, set the default `pageSize` in https://github.com/openedx/frontend-app-course-authoring/pull/1050/files#diff-8eef2c5f738aa3151262ca0ca6634eb16c533410f9286ce96c0a58061590bac4R57 to a lower number, and confirm going through multiple pages of results)
    1. The route in the URL changes to `/libraries`
    1. When refreshing the page, it should keep you on the "Libraries" tab 
    1. When clicking on a v2 library it takes you to the library authoring mfe for that v2 library
    1. When clicking on "New Library" button it should take you to the library authoring mfe to create a new v2 library
    1. When disabling the `studio.library_authoring_mfe` waffle flag, and then clicking on a v2 library it should take you to the placeholder page (shown in screenshot above), the url should reflect the clicked v2 library (eg: `/library/lib:SampleTaxonomyOrg1:AL1`)
    1. The same should happen when clicking on "New Library" when `studio.library_authoring_mfe` waffle flag is disabled, it should redirect you to the placeholder page
1. When clicking on the "Legacy Libraries" tab:
    1. v1 libraries are shown
    1. The route in the URL changes to `/libraries-v1`
    1. When refreshing the page it should keep you on the "Legacy Libraries" tab 
    1. When clicking on a v1 library it takes you to the legacy studio UI for the clicked library
    1. Since the default `LIBRARY_MODE` is "mixed", clicking on "New Library" should still take you to the library authoring mfe to create a new v2 library
1. Change the `LIBRARY_MODE` [flag](https://github.com/openedx/frontend-app-course-authoring/pull/1050/files#diff-da4c41d59c967338247b2f0ea6d845ab05171f942fd9521b96d3304202185db9R49) to "v2 only" and restart the course-authoring container
    1. Confirm that there is only 1 tab "Library" that shows up
    1. When clicked it shows v2 libraries
    1. The route changes to `/libraries`  
    1. When clicking on "New Library" it takes you to the library authoring mfe if enabled to create new v2 libraries otherwise it takes you to a placeholder page
1. Change the `LIBRARY_MODE` [flag](https://github.com/openedx/frontend-app-course-authoring/pull/1050/files#diff-da4c41d59c967338247b2f0ea6d845ab05171f942fd9521b96d3304202185db9R49) to "v1 only" and restart the course-authoring container
    1. Confirm that there is only 1 tab "Library" that shows up
    1. When clicked it shows v1 libraries
    1.  The route changes to `/libraries-v1`
    1. When clicking "New Library" it takes you to the legacy UI to create a new v1 library
---
Private-ref: [FAL-3751](https://tasks.opencraft.com/browse/FAL-3751)

openedx-webhooks commented:
Thanks for the pull request, @yusuf-musleh! Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.

<!-- comment:end_of_wip -->




<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IHRydWV9 -->

bradenmacdonald commented on src/studio-home/tabs-section/index.jsx:
```diff
@@ -23,9 +25,14 @@ const TabsSection = ({
 }) => {
   const navigate = useNavigate();
+
+  // TODO: this should be a flag in the backend
```
This can just be [a frontend config variable](https://github.com/openedx/frontend-app-course-authoring/blob/master/.env). I'm not expecting any backend changes for this PR (no waffle flags). Sorry if that wasn't clear.

bradenmacdonald commented on src/studio-home/data/api.js:
```diff
@@ -40,6 +40,11 @@ export async function getStudioHomeLibraries() {
   return camelCaseObject(data);
 }
 
+export async function getStudioHomeLibrariesV2() {
```
This function should define the type of the returned data.

yusuf-musleh replied:
Will do.

bradenmacdonald commented:
@yusuf-musleh Can you check if there's a way to update the URL, so when I click on the "Libraries" tab, the URL changes from `/home` to `/libraries` or `/home?tab=libraries`, and I can bookmark and return to that view?

yusuf-musleh commented:
> Can you check if there's a way to update the URL, so when I click on the "Libraries" tab, the URL changes from /home to /libraries or /home?tab=libraries, and I can bookmark and return to that view?

@bradenmacdonald  I've implemented this here in [17a5070](https://github.com/openedx/frontend-app-course-authoring/pull/1050/commits/17a507008137e74af4c1d4e153591f8fa19ebe85). The url path now updates with `/libraries` or `/legacy-libraries` and `/home` when the tabs change. They also open the respective tab when being accessed with the path directly and supports the browsers back/forward keys. 

I opted for paths rather than query params as react-router would always refresh the page when dealing/changing query params (I've tried a bunch of different approaches), as opposed to changing the paths when the routes point to the same component which doesn't cause a refresh.

yusuf-musleh commented:
@bradenmacdonald  I had to temporarily remove the ts/tsx related code to get the tests to run ([da189c1](https://github.com/openedx/frontend-app-course-authoring/pull/1050/commits/da189c1d6d1687b5a6971ab8604fd852481c4a6b) and [14933d2](https://github.com/openedx/frontend-app-course-authoring/pull/1050/commits/14933d2ce093c4396469c62b9e307275940fd947)). I assume once the new frontend-build is merged this will no longer be an issue.

bradenmacdonald commented:
@yusuf-musleh  Yes, that's right.

rpenido commented:
Hi @yusuf-musleh! Just to note that I think you should skip the tests from `LibraryV2Placeholder.jsx` (using `/* istanbul ignore next */`) as this file will be removed in the PR that adds the Library Home page.

yusuf-musleh commented:
@rpenido Good point! Thanks for the pointer, I've added it.

pomegranited commented on src/studio-home/tabs-section/libraries-v2-tab/index.jsx:
```diff
@@ -0,0 +1,108 @@
+            <CardItem
+              key={`${org}+${slug}`}
+              isLibraries
+              displayName={title}
```
The [ContentLibrary model](https://github.com/openedx/edx-platform/blob/51ebabfd96b1d3c0d096c46981cc09e17ef11388/openedx/core/djangoapps/content_libraries/models.py#L84) doesn't have a `title` field.. I'm not sure what should go here instead?

yusuf-musleh replied:
That's right, however the [serializers](https://github.com/openedx/edx-platform/blob/master/openedx/core/djangoapps/content_libraries/serializers.py#L23-L61) have the `title` field.

Here is an example of what is returned in the from the API:

```js
    {
        "id": "lib:SampleTaxonomyOrg1:TL1",
        "type": "complex",
        "org": "SampleTaxonomyOrg1",
        "slug": "TL1",
        "title": "Test Library 1",
        "description": "",
        "num_blocks": 0,
        "version": 0,
        "last_published": null,
        "allow_lti": false,
        "allow_public_learning": false,
        "allow_public_read": false,
        "has_unpublished_changes": false,
        "has_unpublished_deletes": false,
        "license": ""
    }
```

pomegranited replied:
Oh my mistake, you're totally correct, there is a `title` field in the serializer, and it's [the library's learning package title](https://github.com/openedx/edx-platform/blob/master/openedx/core/djangoapps/content_libraries/api.py#L284). But since I was naughty and neglected to use the `content_libraries.api` to create my libraries, they didn't have learning packages.

My bad!

yusuf-musleh replied:
ohh I see, gotcha!

pomegranited commented on src/studio-home/__mocks__/studioHomeMock.js:
```diff
@@ -62,7 +62,7 @@ module.exports = {
   ],
   librariesEnabled: true,
-  libraryAuthoringMfeUrl: 'http://localhost:3001',
+  libraryAuthoringMfeUrl: 'http://localhost:3001/',
```
The platform default for this URL [does not contain the trailing slash](https://github.com/openedx/edx-platform/blob/51ebabfd96b1d3c0d096c46981cc09e17ef11388/cms/envs/devstack.py#L178), but usually with a fully-deployed MFE, there's a `PUBLIC_PATH` variable that can default to `/`.

But I don't know for sure that our MFE configurations are consistent enough to rely on the presence of a trailing slash. Instead of constructing Library Authoring URLs from simple string concatenation, could you create a function that handles inserting the trailing slash if needed?

yusuf-musleh replied:
Interesting, I wasn't familiar with that! That's a good idea, I've added the function here https://github.com/openedx/frontend-app-course-authoring/pull/1050/commits/567dcb48f08d7bb43ea389bda9ee3dd9c1c22966

pomegranited reviewed (changes requested):
Hi @yusuf-musleh , I found a few things to fix, but this is working fine in general.

Feel free to push back on anything :)

yusuf-musleh commented on src/studio-home/StudioHome.test.jsx:
```diff
@@ -145,7 +158,18 @@ describe('<StudioHome />', async () => {
+        setConfig({
+          ...getConfig(),
+          LIBRARY_MODE: 'v1 only',
+        });
```
For this test, the library mode needs to be set to `v1 only`. I set it back to `mixed` inside `beforeEach` to make sure the rest of the tests have the correct mode.

pomegranited replied:
Oh gotcha.. thanks for explaining!

yusuf-musleh commented on src/studio-home/__mocks__/studioHomeMock.js:
```diff
@@ -62,7 +62,7 @@ module.exports = {
     },
   ],
   librariesEnabled: true,
```
I believe this scenarios will be checked as part of @rpenido 's [PR](https://github.com/open-craft/frontend-app-course-authoring/pull/37/files), since the placeholder is going to be swapped out to an actual page.

yusuf-musleh commented on src/studio-home/StudioHome.jsx:
```diff
@@ -79,8 +82,10 @@ const StudioHome = ({ intl }) => {
+    if (isMixedOrV2LibrariesMode(libMode)) {
+      libraryHref = libraryAuthoringMfeUrl && redirectToLibraryAuthoringMfe
+        ? `${libraryAuthoringMfeUrl}create`
+        : `${window.location.origin}/course-authoring/library/create`;
```
Got it, updated it here: https://github.com/openedx/frontend-app-course-authoring/pull/1050/commits/28597411042e5c2fadf47cc8c7ef669725a4405f

yusuf-musleh commented on src/studio-home/tabs-section/libraries-v2-tab/index.jsx:
```diff
@@ -0,0 +1,108 @@
+      ? `${libraryAuthoringMfeUrl}library/${id}`
+      // Redirection to the placeholder is done in the MFE rather than
+      // through the backend i.e. redirection from cms, because this this will probably change
+      : `${window.location.origin}/course-authoring/library/${id}`
```
Updated: https://github.com/openedx/frontend-app-course-authoring/pull/1050/commits/28597411042e5c2fadf47cc8c7ef669725a4405f

yusuf-musleh commented on src/studio-home/tabs-section/TabsSection.test.jsx:
```diff
@@ -82,9 +131,45 @@ describe('<TabsSection />', () => {
+    axiosMock.onGet(getStudioHomeApiUrl()).reply(200, data);
+    await executeThunk(fetchStudioHomeData(), store.dispatch);
+
+    expect(screen.getByText(tabMessages.librariesTabTitle.defaultMessage)).toBeInTheDocument();
```
Good point, added here: https://github.com/openedx/frontend-app-course-authoring/pull/1050/commits/beda37f829cbdd43204144ecd2f0217d09d8f669

yusuf-musleh commented on src/index.jsx:
```diff
@@ -52,6 +53,9 @@ const App = () => {
       <Route>
         <Route path="/home" element={<StudioHome />} />
+        <Route path="/libraries" element={<StudioHome />} />
+        <Route path="/legacy-libraries" element={<StudioHome />} />
```
I agree, it's cleaner that way, updated here: [a0a30b7](https://github.com/openedx/frontend-app-course-authoring/pull/1050/commits/a0a30b7f31dd3ad9a0f518ac0932143ea81b0e14)

yusuf-musleh commented on src/studio-home/tabs-section/index.jsx:
```diff
@@ -118,10 +172,17 @@ const TabsSection = ({
+      navigate(
+        libMode === 'v1 only'
+          ? '/libraries'
+          : '/legacy-libraries',
```
That makes sense, I've changed it here: https://github.com/openedx/frontend-app-course-authoring/pull/1050/commits/094086e05f948cecb3437f89026a959ce0ff45a7

yusuf-musleh commented:
@pomegranited Thanks for the detailed review! I believe I addressed all the comments and it should be ready to go for another round when you get a chance. 

One more thing to note is, once the new frontend-build PRs are merged, I need to revert some commits on this PR to bring back the typescript code I removed to get the tests running:
- [temp: This removes TS code to get tests to run](https://github.com/openedx/frontend-app-course-authoring/pull/1050/commits/efbc625aaa03a51e0f010ae25a7821b0404b854a)
- [temp: Rename .tsx -> .jsx & .ts -> .js for tests](https://github.com/openedx/frontend-app-course-authoring/pull/1050/commits/79e6516b32cc6b03e5438f5c33138a20d72bea24)

pomegranited reviewed (approved):
:+1: Works perfectly, and the code looks great.

- [x] I tested this using my tutor dev stack and @yusuf-musleh 's spectacular testing instructions :)
- [x] I read through the code
- [x] I checked for accessibility issues by using my keyboard alone to navigate the new tabs and pages.
- [x] Includes documentation -- updates README
- [x] User-facing strings are extracted for translation

yusuf-musleh commented:
Thanks @pomegranited for the great review!

@xitij2000 This is ready for CC when you have a chance.

bradenmacdonald commented:
Before we merge this, please make 100% sure that in production (on edx.org) it will continue to display the status quo (v1 only), and for developers show both.

yusuf-musleh commented:
> Before we merge this, please make 100% sure that in production (on edx.org) it will continue to display the status quo (v1 only), and for developers show both.

@bradenmacdonald  The default `LIBRARY_MODE` is set to `"v1 only"` in `.env` for [production](https://github.com/openedx/frontend-app-course-authoring/pull/1050/files#diff-e9cbb0224c4a3d23a6019ba557e0cd568c1ad5e1582ff1e335fb7d99b7a1055dR46) and `"mixed"` in the `.env.development` for [developers](https://github.com/openedx/frontend-app-course-authoring/pull/1050/files#diff-da4c41d59c967338247b2f0ea6d845ab05171f942fd9521b96d3304202185db9R49). 

We also have a [test](https://github.com/openedx/frontend-app-course-authoring/pull/1050/files#diff-1dd24886b21ffc32f7be0703db4ccb8fc6937fa9408c7d7c6115f5921d623fefR139-R158) to confirm that the v1 "legacy" libraries only show when in `"v1 only"` mode.

The only difference to the current status quo is when clicking on the "Libraries" tab the url will change to `/libraries-v1`. Previously the URL did not change at all when changing tabs.

xitij2000 commented:
@yusuf-musleh  Sorry for the delay here but I have a completely broken devstack and needed to start from scratch. I will try to review this soon, but do ping me if you'd like it reviewed sooner.

bradenmacdonald commented:
@yusuf-musleh If the frontend-build PR doesn't get merged by the time this PR is ready to merge (hopefully Monday, Tuesday at the latest), feel free to merge it with the JS version and we can do a fast-follow to convert it to TypeScript as soon as the frontend-build PR does get merged. I want to try and unblock us from this big pile of PRs.

bradenmacdonald commented:
> I have a completely broken devstack and needed to start from scratch

@xitij2000  Since @pomegranited has already tested it, you don't necessarily need to run the code yourself as part of your CC review - reading it and reviewing it that way would be completely fine. Of course, if you feel that it's important for you to test it as well on your own devstack, or that's useful for your review process, that's probably even better and we can wait a day or two.

bradenmacdonald commented:
@xitij2000 @yusuf-musleh Also, please be aware that you can add the `create-sandbox` label to a PR like this and Grove will spin up a sandbox where you can test the feature. Though in this case it would probably use the "production" settings and you would only see the v1 tab.

xitij2000 commented:
> @xitij2000 @yusuf-musleh Also, please be aware that you can add the `create-sandbox` label to a PR like this and Grove will spin up a sandbox where you can test the feature. Though in this case it would probably use the "production" settings and you would only see the v1 tab.

I generally prefer to run and test the code as well especially after a review round in case something breaks. My devstack is fixed now so I am reviewing this right now.

yusuf-musleh commented:
@bradenmacdonald 

> feel free to merge it with the JS version and we can do a fast-follow to convert it to TypeScript as soon as the frontend-build PR does get merged.

Sounds good, will do.

yusuf-musleh commented:
> My devstack is fixed now so I am reviewing this right now.

@xitij2000 Just checking in to see in case I missed the review or if it wasn't submitted yet.

xitij2000 reviewed (approved):
Looks good! Just feel like the image doesn't belong  here.

yusuf-musleh commented on README.rst:
```diff
@@ -264,6 +264,22 @@ In additional to the standard settings, the following local configuration items
+=================================
+
+.. image:: ./docs/readme-images/feature-v2-and-legacy-libs.png
+
```
Thanks for the review! Good point, I've removed it.

openedx-webhooks commented:
@yusuf-musleh 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/frontend-app-course-authoring&entry.867055334=https://github.com/openedx/frontend-app-course-authoring/pull/1050&entry.1484655318=https://github.com/yusuf-musleh&entry.752974735=2024-05-27+17:20&entry.1917517419=2024-06-20+12:00&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->
